pretix_wallet/payment.py

```diff
@@ -73,31 +77,25 @@ def _get_payment_info_data(self, wallet: CustomerWallet):
             'retry': True
         }
 
+
+
+
     def execute_payment(self, request: HttpRequest, payment: OrderPayment, is_early_special_case=False) -> str:
         # re-implemented as the original method does not allow for giftcards to have negative balance
 
         gcpk = payment.info_data.get('gift_card')
         if not gcpk:
             raise PaymentException("Invalid state, should never occur.")
         try:
-            with transaction.atomic():
-                try:
-                    gc = GiftCard.objects.select_for_update(of=OF_SELF).get(pk=gcpk)
-                except GiftCard.DoesNotExist:
-                    raise PaymentException(_("This gift card does not support this currency."))
-                if gc.currency != self.event.currency:  # noqa - just a safeguard
-                    raise PaymentException(_("This gift card does not support this currency."))
-                if not gc.accepted_by(self.event.organizer):
-                    raise PaymentException(_("This gift card is not accepted by this event organizer."))
-
-                trans = gc.transactions.create(
-                    value=-1 * payment.amount,
-                    order=payment.order,
-                    payment=payment,
-                    acceptor=self.event.organizer,
-                )
-                payment.info_data['transaction_id'] = trans.pk
-                payment.confirm(send_mail=not is_early_special_case, generate_invoice=not is_early_special_case)
+            try:
+                gc = GiftCard.objects.select_for_update(of=OF_SELF).get(pk=gcpk)
```

Comment:
.get() evaluates the queryset, which activates the lock. if you want to keep it locked, the evaluation needs to happen inside the same transaction as the other operations

Reply:
is there a reason to keep the lock? Shouldn't the lock around the write part be enough? Ore are you worried about the gift card changing owners?